### PR TITLE
Port TG's chat subsystem

### PR DESCRIPTION
**Pull request by Couls**
**What does this PR do:**
Port TG's new chat subsystem https://github.com/tgstation/tgstation/pull/45354 which they ported from TGMC https://github.com/tgstation/TerraGov-Marine-Corps/pull/1043

The chat subsystem queues new chat messages and prevents lag spikes that usually happened when you'd be in a situation where you'd get a lot of messages at once, this means fastmos no longer KILLS your game, at least in my tests with 5 monkeys and 15 glass shards and deleting the floor, the game didn't even hang and this was on my laptop. Big performance improvement in that regard

Shotguns were also a culprit of this lag, shotgun buff or maybe nerf? depends I guess

**Changelog:**
:cl:
add: new chatsubsystem from TG!
/:cl:

### CODE REVIEW

**Comment by Terilia** [comment]
Sounds nice.

**Comment by AffectedArc07** [comment]
```
Shotguns were also a culprit of this lag
```

Sounds about right tbh

File: code/controllers/subsystem/chat.dm [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+	//url_encode it TWICE, this way any UTF-8 characters are able to be decoded by the Javascript.
+	//Do the double-encoding here to save nanoseconds
+	var/twiceEncoded = url_encode(url_encode(message))
+
+	if(islist(target))
+		for(var/I in target)
+			var/client/C = CLIENT_FROM_VAR(I) //Grab us a client if possible
+
+			if(!C?.chatOutput || C.chatOutput.broken) //A player who hasn't updated his skin file.
```
Comment (KasparoVy) [inline code/controllers/subsystem/chat.dm]: ? FTW

Comment (AffectedArc07) [reply code/controllers/subsystem/chat.dm]: We dont have `chatOutput.broken` do we? I swear the `Fix Chat` is a TG only verb, so is reverting to normal chat if fancy chat fails to load

Comment (Couls) [reply code/controllers/subsystem/chat.dm]: https://github.com/Couls/Paradise/blob/2c85e79224c2258a8d2bbad69aa8a13649055d48/goon/code/datums/browserOutput.dm#L28

**Pull request merged**